tsl/src/continuous_aggs/planner.c

```diff
@@ -0,0 +1,427 @@
+/*
+ * This file and its contents are licensed under the Timescale License.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-TIMESCALE for a copy of the license.
+ */
+#include <postgres.h>
+
+#include <nodes/makefuncs.h>
+#include <nodes/nodeFuncs.h>
+#include <nodes/nodes.h>
+#include <nodes/pg_list.h>
+#include <parser/parse_func.h>
+
+#include "continuous_aggs/common.h"
+#include "planner.h"
+#include "ts_catalog/continuous_aggs_watermark.h"
+
+/*
+ * The watermark function of a CAgg query is embedded into further functions. It
+ * has the following structure:
+ *
+ * 1. a coalesce expression
+ * 2. an optional to timestamp conversion function (date, timestamp, timestamptz)
+ * 3. the actual watermark function
+ *
+ * For example:
+ *        ... COALESCE(to_timestamp(cagg_watermark(59)), XXX) ...
+ *        ... COALESCE(cagg_watermark(59), XXX) ...
+ *
+ * We use the following data structure while walking to the query to analyze the query
+ * and collect references to the needed functions. The data structure contains:
+ *
+ *   (a) values (e.g., function Oids) which are needed to analyze the query
+ *   (b) values that are changed during walking through the query
+ *       (e.g., references to parent functions)
+ *   (c) result data like the watermark functions and their parent functions
+ */
+typedef struct
+{
+	/* (a) Values initialized after creating the context */
+	Oid watermark_function_oid;	  // Oid of the watermark function
```

Comment:
This oid never changes over the lifetime of a backend so we could only look this up once and have the oid stored in a static variable

Reply:
Good idea, thanks!!! Will push and adjustment. Thanks for your revieew!

Reply:
Good catch! Will push an adjustment before merge it. Thanks for the review!!

Reply:
Actually it can change if the user DROP/CREATE the function.
```sql
21737 (leader) fabrizio=# select oid from pg_proc where proname = 'cagg_watermark' and pronamespace = '_timescaledb_functions'::regnamespace;
  oid  
-------
 16758
(1 row)

21737 (leader) fabrizio=# begin ;
BEGIN
Time: 1.155 ms
21737 (leader) fabrizio=*# alter extension timescaledb drop function _timescaledb_functions.cagg_watermark(hypertable_id INTEGER);
ALTER EXTENSION
Time: 2.072 ms
21737 (leader) fabrizio=*# drop function _timescaledb_functions.cagg_watermark(hypertable_id INTEGER) ;
DROP FUNCTION
Time: 3.824 ms
21737 (leader) fabrizio=*# CREATE FUNCTION _timescaledb_functions.cagg_watermark(hypertable_id INTEGER)
21737 (leader) fabrizio-*# RETURNS INT8 AS '$libdir/timescaledb-2.14.0-dev', 'ts_continuous_agg_watermark' LANGUAGE C STABLE STRICT PARALLEL RESTRICTED;
CREATE FUNCTION
Time: 2.736 ms
21737 (leader) fabrizio=*# select oid from pg_proc where proname = 'cagg_watermark' and pronamespace = '_timescaledb_functions'::regnamespace;
  oid  
-------
 17085
(1 row)
```
On cloud it will never happen due to permission restrictions but on premises users can enter in this very corner case.

Reply:
Never mind, it will never happen because the function dependency. Changed it to a static variable!